### DISCUSSION TITLE
Improved type inference with arrays and the new() syntax.

### DISCUSSION THREAD

**Discussion by audacode** [dotnet/csharplang#7175] [Language Ideas] [answered]
When constructing an object graph, it seems out of place not to be able to use new() with type inference when constructing arrays.  Is there a reason the type information cannot be inferred from the left side?  I have large object graphs without any type names appearing, apart from this one exception (surely it would be reasonable to improve the developer experience).

Here are examples that all work, apart from the last one.  Shouldn't the compiler be able to infer the items on the right are of type Student? (see the line producing an error with the comment at the end):

```cs
        Student student = new("John");
        string[] names = new[] { "John", "Jane" };
        Student[] studentsA = new[] { new Student("John"), new Student("Jane") }; }
        Student[] studentsB = new[] { new("John"), new("Jane") }; // Compile error (cannot infer)

        record Student(string Name);
```

Could this be improved in the next version?  Also, please make sure any improvements flow through to the new literal syntax shown below (targeted for C# 12?):
```cs
Student[] students = [new("John"),new("Jane")];
```
Thanks you.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/2701

**Reply by ufcpp** [reply, marked as answer]
https://github.com/dotnet/csharplang/blob/af3e3b47d34b86d239137b2c66ee04c42b5f96cd/meetings/2020/LDM-2020-09-09.md?plain=1#L163

**Reply by audacode** [reply]
Thank you.  Yes, I also just found this myself:
https://github.com/dotnet/csharplang/issues/2701

That thread implies it will work with this syntax:
```cs
Student[] students = [new("John"),new("Jane")];
```

But probably not with the older syntax.  I still think it should be a goal in terms of language consistency,

**Comment by audacode** [comment]
I am closing given there is an open issue tracking this (link above)
